File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    int n;
    
    bool isPalindrome(string s, int l, int r){
        while(l < r){
            if(s[l] != s[r])
                return false;
            l++;
            r--;
        }
        return true;
    }
    
    void solve(string s, int ind, vector<string> &temp, vector<vector<string>> &ans){
        
        if(ind >= s.size()){
            ans.push_back(temp);
            return;
        }

        for(int i = ind; i < s.size(); i++){
            if(isPalindrome(s,ind,i)){
                // push the substring from ind to i as we have partitioned till i
                temp.push_back(s.substr(ind, i - ind + 1));
                solve(s, i + 1 , temp, ans);
                temp.pop_back();
            }
        }
    }
    
    vector<vector<string>> partition(string s) {
        
        n = s.size();
        vector<vector<string>> ans;
        vector<string> temp ;
        
        solve(s, 0, temp, ans);
        return ans;
        
        
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();
        solve(s, 0, new ArrayList<>(), ans);
        return ans;   
    }

    public boolean isPalindrome(String s){
        int left = 0, right = s.length() - 1;
        while(left < right){
            if(s.charAt(left) != s.charAt(right)){
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    public void solve(String s, int start, List<String> temp, List<List<String>> ans){

        if(start >= s.length()){
            ans.add(new ArrayList<>(temp));
            return;
        }

        for(int i = start; i < s.length(); i++){
            String subStr = s.substring(start, i + 1);
            if(isPalindrome(subStr)){
                temp.add(subStr);
                solve(s, i + 1, temp, ans);
                temp.remove(temp.size() - 1);
            }
        }
    }
}